File: core/services/user/feedback.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import List, Tuple

from services.search import features
from services.story import story, source, author
from services.user import favorite, read_story
from services import mongo
from shared import bson_id
from shared.types import data_set_types

DB_COLLECTION_NAME = "UserFeedback"
USER_FEEDBACK_COUNT = 200
RANKING_DATA_FAVORITE_COUNT = 50
FEEDBACK_RECEIVED_REWARD = 0.1
URL_CLICKED_SCORE = 1.0
SHARED_SCORE = 5.0
ANGRY_SCORE = -5.0
CRY_SCORE = -2.0
NEUTRAL_SCORE = 0.0
SMILE_SCORE = 2.0
HAPPY_SCORE = 5.0


class UserFeedback(BaseModel):
  id: bson_id.ObjectIdStr = Field(None, alias='_id')
  user_id: str
  story_id: str
  search_term: str = "*"
  feedback_datetime: datetime = Field(default_factory=datetime.utcnow)
  feedback_type: int


def add(user_feedback: UserFeedback):
  """Straight adding of a UserFeedback object to Mongo.

    Args:
        user_feedback: The user's feedback to be added to Mongo.

    Returns:
        The results of the mongo operation.

    """

  return mongo.add_or_update(user_feedback.dict(), DB_COLLECTION_NAME)


def remove_feedback_of_user(user_id: str):
  """Remove all feedbacks for a given user from Mongo.

    Args:
        user_id: The user id for which to remove the feedback.

    Returns:
        The amount of user feedbacks deleted.

    """

  mongo_filter = {"user_id": user_id}
  r = mongo.remove(DB_COLLECTION_NAME, mongo_filter)
  return r.deleted_count


def get_list(user_id: str) -> Tuple[UserFeedback, ...]:
  """Retrieve a list of feedbacks from a user

    Args:
        user_id: The user id for which to remove the feedback.

    Returns:
        A tuple of user feedbacks.

    """

  mongo_filter = {}
  if user_id != 'defaultmodel':
    mongo_filter.update({"user_id": user_id})
  feedbacks = mongo.get(DB_COLLECTION_NAME,
                        mongo_filter,
                        limit=USER_FEEDBACK_COUNT,
                        sort="feedback_datetime",
                        reverse=True)
  return tuple(UserFeedback(**feedback) for feedback in feedbacks)


def convert_feedback_type_to_relevancy_rate(feedback_type: int) -> float:
  if feedback_type == 0:
    return URL_CLICKED_SCORE
  elif feedback_type == 1:
    return SHARED_SCORE
  elif feedback_type == 31:
    return ANGRY_SCORE
  elif feedback_type == 32:
    return CRY_SCORE
  elif feedback_type == 33:
    return NEUTRAL_SCORE
  elif feedback_type == 34:
    return SMILE_SCORE
  elif feedback_type == 35:
    return HAPPY_SCORE


def get_tf_training_data(user_id: str) -> List[data_set_types.RankingData]:
  """Retrieves a list of tensorflow training data generated from the user
  feedback.

    Args:
        user_id: The user id for which to retrieve the training data.

    Returns:
        A list of tensorflow training data.

    """

  feedback_list = get_list(user_id)
  data_entry_list = []
  relevancy_dict = {}
  time_dict = {}
  for feedback in feedback_list:
    relevancy_rate = convert_feedback_type_to_relevancy_rate(
        feedback.feedback_type)
    if feedback.story_id in relevancy_dict:
      relevancy_dict[feedback.story_id] += relevancy_rate
    else:
      relevancy_dict.update({feedback.story_id: relevancy_rate})
    if feedback.story_id in time_dict:
      if feedback.feedback_datetime > time_dict[feedback.story_id]:
        time_dict.update({feedback.story_id: feedback.feedback_datetime})
    else:
      time_dict.update({feedback.story_id: feedback.feedback_datetime})
  for story_id in relevancy_dict.keys():
    current_story = story.get_by_id(story_id)
    if current_story:
      ranking_features = features.extract_ranking_features(current_story)
      if ranking_features:
        data_entry = {
            "story_id": story_id,
            "user_id": user_id,
            "relevancy_rate": relevancy_dict[story_id],
            "time_stamp": datetime.timestamp(time_dict[story_id])
        }
        data_entry.update(ranking_features.dict())
        data_entry_list.append(data_set_types.RankingData(**data_entry))
  return data_entry_list


def received(feedback: UserFeedback):
  """Updates relevant values after receiving feedback on a news story from a
  user. Stories which have been read by the user are accounted for and the
  relevancy rate of the user's favorites are updated. The overall reputation
  of the source and author are also updated.

    Args:
        feedback: The feedback received from the User.

    Returns:
        The result of adding the feedback to mongo.

    """

  current_read_story = read_story.ReadStory(story_id=feedback.story_id,
                                            user_id=feedback.user_id)
  read_story.add(current_read_story)
  result = add(feedback)
  reward = None
  #Angry
  if feedback.feedback_type == 31:
    reward = -FEEDBACK_RECEIVED_REWARD
  #Happy or Shared
  elif feedback.feedback_type == 35 or feedback.feedback_type == 1:
    reward = FEEDBACK_RECEIVED_REWARD
  current_story = story.get_by_id(feedback.story_id)
  if current_story:
    story.update_feedback_counts(feedback.story_id, feedback.feedback_type)
    if reward:
      if current_story.keywords:
        for keyword in current_story.keywords:
          favorite.update_from_story(
              feedback.user_id,
              keyword.keyword.text,
              keyword.keyword.text,
              reward,
              favorite.FAVORITE_KEYWORD_DB_COLLECTION_NAME,
              language=current_story.language)
      if current_story.entities:
        for e in current_story.entities:
          favorite.update_from_story(
              feedback.user_id,
              e.entity.links,
              e.entity.text,
              reward,
              favorite.FAVORITE_ENTITY_DB_COLLECTION_NAME,
              language=current_story.language)
      if current_story.source:
        favorite.update_from_story(feedback.user_id,
                                   current_story.source.source_id,
                                   current_story.source.name,
                                   reward,
                                   favorite.FAVORITE_SOURCE_DB_COLLECTION_NAME,
                                   language=current_story.language)
        source.update_reputation(current_story.source.source_id, reward)
      if current_story.author:
        favorite.update_from_story(feedback.user_id,
                                   current_story.author.author_id,
                                   current_story.author.name,
                                   reward,
                                   favorite.FAVORITE_AUTHOR_DB_COLLECTION_NAME,
                                   language=current_story.language)
        author.update_reputation(current_story.author.author_id, reward)
  return result

File: core/internal/feedback/feedback.go
package feedback

import (
	"time"

	"core/internal/dbs"
	"core/internal/story"
	"core/internal/user"
	"core/internal/utils"

	"github.com/google/uuid"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

const userFeedbackCollection = "UserFeedback"
const USER_FEEDBACK_COUNT = 200
const RANKING_DATA_FAVORITE_COUNT = 50
const FEEDBACK_RECEIVED_REWARD = 0.1
const URL_CLICKED_SCORE = 1.0
const SHARED_SCORE = 5.0
const ANGRY_SCORE = -5.0
const CRY_SCORE = -2.0
const NEUTRAL_SCORE = 0.0
const SMILE_SCORE = 2.0
const HAPPY_SCORE = 5.0

type FeedbackType string

const (
	READ    FeedbackType = "read"
	SHARED  FeedbackType = "shared"
	ANGRY   FeedbackType = "angry"
	CRY     FeedbackType = "cry"
	NEUTRAL FeedbackType = "neutral"
	SMILE   FeedbackType = "smile"
	HAPPY   FeedbackType = "happy"
)

type UserFeedback struct {
	UserId           string       `bson:"user_id,omitempty" json:"user_id,omitempty"`
	StoryId          string       `bson:"story_id,omitempty" json:"story_id,omitempty"`
	SearchTerm       string       `bson:"search_term,omitempty" json:"search_term,omitempty"`
	FeedbackDatetime time.Time    `bson:"feedback_datetime,omitempty" json:"feedback_datetime,omitempty"`
	FeedbackType     FeedbackType `bson:"feedback_type,omitempty" json:"feedback_type,omitempty"`
}

func RemoveOldFeedback() int64 {
	mongoFilter := bson.M{
		"feedback_datetime": bson.M{
			"$lte": primitive.NewDateTimeFromTime(time.Now().AddDate(0, 0, -story.STORY_DAYS_TO_EXPIRY)),
		},
	}
	return dbs.Remove(userFeedbackCollection, mongoFilter)
}

func GetFeedbackList() ([]UserFeedback, error) {
	mongoFilter := bson.M{
		"feedback_datetime": bson.M{
			"$gte": primitive.NewDateTimeFromTime(time.Now().AddDate(0, 0, -story.STORY_DAYS_TO_EXPIRY)),
			"$lt":  primitive.NewDateTimeFromTime(time.Now()),
		},
	}
	var feedbacks []UserFeedback
	err := dbs.GetSorted(userFeedbackCollection,
		mongoFilter,
		&feedbacks,
		"feedback_datetime",
		true,
		USER_FEEDBACK_COUNT)
	return feedbacks, err
}

func ConvertFeedbackTypeToRelevancyRate(feedbackType FeedbackType) float32 {
	switch feedbackType {
	case READ:
		return URL_CLICKED_SCORE
	case SHARED:
		return SHARED_SCORE
	case ANGRY:
		return ANGRY_SCORE
	case CRY:
		return CRY_SCORE
	case NEUTRAL:
		return NEUTRAL_SCORE
	case SMILE:
		return SMILE_SCORE
	case HAPPY:
		return HAPPY_SCORE
	default:
		return 0
	}
}

func FeedbackReceived(userFeedback UserFeedback) error {
	readStory := user.ReadStory{
		UserId:   userFeedback.UserId,
		StoryId:  userFeedback.StoryId,
		ReadTime: time.Now(),
	}
	user.AddReadStory(readStory)
	dbs.AddOrUpdateOne(userFeedbackCollection, userFeedback)
	var reward float32
	if userFeedback.FeedbackType == ANGRY {
		reward = -FEEDBACK_RECEIVED_REWARD
	}
	if userFeedback.FeedbackType == HAPPY || userFeedback.FeedbackType == SHARED {
		reward = FEEDBACK_RECEIVED_REWARD
	}
	currentStory, er := story.GetStoryById(userFeedback.StoryId)
	if er != nil {
		return utils.LogError(er)
	}
	uid, err := uuid.Parse(userFeedback.StoryId)
	if err != nil {
		return utils.LogError(err)
	}
	story.UpdateFeedbackCounts(uid, ConvertFeedbackTypeToString(userFeedback.FeedbackType))
	if reward == 0 {
		return nil
	}
	for _, keyword := range *currentStory.Keywords {
		user.UpdateFavoriteFromStory(userFeedback.UserId, *keyword.Keyword.Text, *keyword.Keyword.Text, reward, user.FAVORITE_KEYWORD_DB_COLLECTION_NAME, *currentStory.Language)
	}
	for _, entity := range *currentStory.Entities {
		user.UpdateFavoriteFromStory(userFeedback.UserId, *entity.Entity.Links, *entity.Entity.Text, reward, user.FAVORITE_ENTITY_DB_COLLECTION_NAME, *currentStory.Language)
	}
	user.UpdateFavoriteFromStory(userFeedback.UserId, currentStory.Source.SourceId, *currentStory.Source.Name, reward, user.FAVORITE_SOURCE_DB_COLLECTION_NAME, *currentStory.Language)
	story.UpdateSourceReputation(currentStory.Source.SourceId, reward)
	user.UpdateFavoriteFromStory(userFeedback.UserId, currentStory.Author.AuthorId.String(), *currentStory.Author.Name, reward, user.FAVORITE_AUTHOR_DB_COLLECTION_NAME, *currentStory.Language)
	story.UpdateAuthorReputation(currentStory.Author.AuthorId, reward)
	return nil
}

func ConvertFeedbackTypeToString(feedbackType FeedbackType) string {
	switch feedbackType {
	case READ:
		return "read"
	case SHARED:
		return "shared"
	case ANGRY:
		return "angry"
	case CRY:
		return "cry"
	case NEUTRAL:
		return "neutral"
	case SMILE:
		return "smile"
	case HAPPY:
		return "happy"
	default:
		return "s"
	}
}
